documentation/Calls.md

```diff
@@ -0,0 +1,636 @@
+
+Derived type components that are `ALLOCATABLE` are finalized
+and deallocated.
+Components with initializers are (re)initialized.
+
+The preparation of effective arguments for `INTENT(OUT)` could be
+done on either side of the call.  If the preparation is
+done by the caller, there is an optimization opportunity
+in situations where unmodified incoming `INTENT(OUT)` dummy
+arguments are being passed onward as outgoing `INTENT(OUT)`
+arguments.
+
+### Arguments and function results requiring descriptors
+
+Dummy arguments are represented with the addresses of descriptors
+when they have any of the following characteristics:
+
+1. assumed-shape array (`DIMENSION(:)`)
+1. assumed-rank array (`DIMENSION(..)`)
+1. parameterized derived type with assumed `LEN` parameters
+1. polymorphic (`CLASS(T)`, `CLASS(*)`)
+1. assumed-type (`TYPE(*)`)
+1. coarray dummy argument
+1. `INTENT(IN) POINTER` argument (15.5.2.7, C.10.4)
+
+Non-F77ish procedures use descriptors to represent two further
+kinds of dummy arguments:
+
+1. assumed-length `CHARACTER(*)`
+1. dummy procedures
+
+F77ish procedures use other means to convey character length and host instance
+links (respectively) for these arguments.
+
+Function results are described by the caller & callee in
+a caller-supplied descriptor when they have any of the following
+characteristics, some which necessitate an explicit interface:
+
+1. deferred-shape array (so `ALLOCATABLE` or `POINTER`)
+1. derived type with any non-constant `LEN` parameter
+   (C795 prohibit assumed lengths)
+1. procedure pointer result (when the interface must be explicit)
+
+Storage for a function call's result is allocated by the caller when
+possible: the result is neither `ALLOCATABLE` nor `POINTER`,
+the shape is scalar or explicit, and the type has `LEN` parameters
+that are constant expressions.
+In other words, the result doesn't require the use of a descriptor
+but can't be returned in registers.
+This allows a function result to be written directly into a local
+variable or temporary when it is safe to treat the variable as if
+it were an additional `INTENT(OUT)` argument.
+(Storage for `CHARACTER` results, assumed or explicit, is always
+allocated by the caller, and the length is always passed so that
+an assumed-length external function will work when eventually
+called from a scope that declares the length that it will use
+(15.5.2.9 (2)).)
+
+Note that the lower bounds of the dimensions of non-`POINTER`
+non-`ALLOCATABLE` dummy argument arrays are determined by the
+callee, not the caller.
+(A Fortran pitfall: declaring `A(0:9)`, passing it to a dummy
+array `D(:)`, and assuming that `LBOUND(D,1)` will be zero
+in the callee.)
+If the declaration of an assumed-shape dummy argument array
+contains an explicit lower bound expression (R819), its value
+needs to be computed by the callee;
+it may be captured and saved in the incoming descriptor
+as long as we assume that argument descriptors can be modified
+by callees.
```

Comment:
Assuming so would prevent forwarding descriptors when possible, right ?
I would have stored local lower bounds locally to avoid leaking them and having weird bugs, on the other hand if we already have storage space for them, that may be a waste.

Reply:
OK, you already replied to this comment in another comment (in "Before the call" l158. )
You only mention assumed-shape, but for explicit-shape/assumed-sized PDT or polymorphic argument we would also have a descriptor, right ? Will descriptors also be used to store the lower/upper bounds that maybe be non-constants in such cases ?

Reply:
Yes, there are multiple reasons for having descriptors, and if they're around, we might as well use their storage for bounds if necessary.